[FINAL] Added C++ exception handling.

kstenerud opened the pull request:
https://mindsnacks.atlassian.net/browse/CHIM-153

Yes, we now support C++ exceptions!
If an exception is thrown, KSCrash will record what kind of exception it was, and for exceptions that subclass std::exception, it will record the result of exception->what(). For all primitive types (int, float, char*) it will record the raw values (or strings).

With the following code:

```
class MyException: public std::exception
{
public:
    virtual const char* what() const noexcept;
};

const char* MyException::what() const noexcept
{
    return "Something bad happened...";
}

- (void) throwUncaughtCPPException
{
    throw MyException();
}
```

An unhandled C++ exception crash will look like this:

```
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x00000000 at 0x00000000
Crashed Thread:  0

Application Specific Information:
*** Terminating app due to uncaught exception 'MyException', reason: 'Something bad happened...'

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   Crash-Tester                    0x0008e32b 0x85000 + 37675 (-[Crasher throwUncaughtCPPException] + 87)
1   Crash-Tester                    0x0008c10b 0x85000 + 28939 (__32-[AppDelegate(UI) crashCommands]_block_invoke343 + 75)
2   Crash-Tester                    0x0008ccfb 0x85000 + 31995 (-[CommandEntry executeWithViewController:] + 59)
3   Crash-Tester                    0x0008d3a9 0x85000 + 33705 (-[CommandTVC tableView:didSelectRowAtIndexPath:] + 193)
4   UIKit                           0x38ee626d 0x38e21000 + 807533 (<redacted> + 877)
...
```

Review:
- [x] @JaviSoto 
- [x] @tedw4rd

kstenerud committed:
Added C++ exception handling.

JaviSoto commented on Example-Apps/Crash-Tester/Crasher.mm:
```diff
@@ -229,4 +235,9 @@ - (void) userDefinedCrash
 
+- (void) throwUncaughtCPPException
+{
+    throw 20;
```
you should've totally done

``` cpp
int up = 20;
throw up;
```

JaviSoto commented on KSCrash/KSCrash/KSCrashSentry_CPPException.cpp:
```diff
@@ -0,0 +1,159 @@
+        orig_cxa_throw = (cxa_throw_type) dlsym(RTLD_NEXT, "__cxa_throw");
+    }
+    orig_cxa_throw(thrown_exception, tinfo, dest);
+}
```
this is so cool

JaviSoto commented:
<img src="http://25.media.tumblr.com/tumblr_lz4u84XJPc1qfu5wgo1_1280.jpg" />

kstenerud committed:
Added basic identification for C++ exceptions.

Now if an exception is thrown, KSCrash will record what kind of exception it was, and for exceptions that subclass std::exception, it will record the result of exception->what().

kstenerud committed:
User may now specify what kinds of crashes to handle.

kstenerud committed:
Cleaned up KSCrash main interface.

Properties "sink" and "onCrash" are now being managed by an installation in a typical setup, so they shouldn't be exposed by default.

"printTraceToStdout" is of limited use now that stdout gets turned off on devices.

kstenerud committed:
Fixed unit tests.

kstenerud committed:
Using crash type names rather than bare integers for logging.

JaviSoto commented:
:shipit:

JaviSoto commented:
Needs re-merge.

kstenerud committed:
Merge from master.

kstenerud committed:
Generate catch code to reduce clutter.

tedw4rd commented:
Yeah, very cool.  :ship:

JaviSoto commented:
Needs re-merge again?

kstenerud committed:
Added support for deviceAndAppHash in iOS 7+.
Also fixed category issue in KSCrashAdvanced.

kstenerud committed:
C++ exception handling was overriding NSException handling.
The C++ exception handler now checks for NSException and defers to the NSException handler in such a case.

kstenerud committed:
Merge branch 'master' into feature/cpp-exception-handling

Conflicts:
	KSCrash/KSCrash/KSCrashAdvanced.h
